File: SRT.py
"""
Shortest remaining time, when a process arrives, before it enters the ready
    queue, if it has a CPU burst time that is less than the remaining time of
    the currently running process, a preemption occurs. When such a preemption
    occurs, the currently running process is added back to the ready queue.

Author
    Zhihao Deng (dengz5@rpi.edu)

"""
import sys
import copy
import math
import time
from util import Action, Process


class SRT:
    def __init__(self, procs: [Process], alpha: int, t_cs: int):
        self.endedQueue = []          # processes in terminated state
        self.readyQueue = []          # processes in ready state
        self.alpha = alpha
        self.t_cs = t_cs
        self.clock = 0                # in ms
        self.btProc = None            # process in CPU including enter and leave state
        self.ptProc = None            # preempting process

        # processes in state other than ready and terminated
        self.actionQueue = copy.deepcopy(procs)

    """
        Core algorithm of SRT implementation.
    """

    def simulate(self) -> (int, int, int, int, int):
        #  print all processes
        for p in self.actionQueue:
            print(
                "Process {:s} [NEW] (arrival time {:d} ms) {:d} CPU bursts (tau {:d}ms)".format(
                    p.name, p.arrival_time, len(p.burst_time), p.tau))
            p.action_enter = p.action_leave = p.arrival_time
        print("time 0ms: Simulator started for SRT [Q <empty>]")

        stop = len(self.actionQueue)
        while len(self.endedQueue) < stop:

            interest = [
                p for p in self.actionQueue
                if self.clock == p.action_enter or self.clock == p.action_leave]
            interest.sort(key=self.__sort_by_action)
            if self.clock == 2714:
                tmp = 0

            for proc in interest:

                if proc.action == Action.enter_CPU and self.clock == proc.action_leave:
                    self.actionQueue.remove(proc)  # pop from actionQueue
                    # process being push into CPU
                    if proc.remain == 0:
                        # next burst slice
                        proc.remain = proc.burst_time[proc.index]

                    if self.ptProc != None:
                        # preemption, self.ptProc is proc
                        self.readyQueue.remove(self.ptProc)
                        self.readyQueue.append(self.btProc)
                        self.readyQueue.sort(key=self.__sort_by_remain)
                        self.ptProc = None
                    proc.action = Action.burst
                    proc.action_enter = self.clock
                    proc.action_leave = self.clock + proc.remain
                    proc.switch_count += 1
                    self.actionQueue.append(proc)
                    self.btProc = proc

                if proc.action == Action.leave_CPU and self.clock == proc.action_leave:
                    self.actionQueue.remove(proc)  # pop from actionQueue
                    # process finished context switch
                    if proc.remain == -1:
                        # terminated
                        proc.action = Action.terninated
                        self.endedQueue.append(proc)
                        self.btProc = None
                    elif proc.remain == 0:
                        # entering I/O time
                        proc.action = Action.block
                        proc.action_enter = self.clock
                        proc.action_leave = self.clock + \
                            proc.block_time[proc.index]
                        self.btProc = None
                        self.actionQueue.append(proc)
                    else:
                        # process being preempted
                        self.btProc = Action.ready
                        self.btProc = self.ptProc
                        self.ptProc = None
                        self.readyQueue.append(proc)
                        self.readyQueue.sort(key=self.__sort_by_remain)

                if proc.action == Action.new:
                    self.actionQueue.remove(proc)  # pop from actionQueue
                    # process arrive
                    proc.action = Action.ready
                    proc.remain = proc.burst_time[0]
                    self.readyQueue.append(proc)
                    self.readyQueue.sort(key=self.__sort_by_remain)
                    print("time {:d}ms: Process {:s} (tau {:d}ms) ".format(
                        self.clock, proc.name, proc.tau), end="")
                    print("arrived; added to ready queue ", end="")
                    self.printReady()

                if proc.action == Action.burst and self.clock == proc.action_enter:
                    print("time {:d}ms: Process {:s} (tau {:d}ms) ".format(
                        self.clock, proc.name, proc.tau), end="")
                    print("started using the CPU with {:d}ms burst remaining ".format(
                        proc.remain), end="")
                    self.printReady()

                    if len(self.readyQueue) > 0:
                        proc2 = self.readyQueue[0]
                        if proc2.tau < proc.tau:
                            # preempting
                            proc.action = Action.leave_CPU
                            proc.action_enter = self.clock
                            proc.action_leave = self.clock + \
                                (self.t_cs >> 1)
                            proc.preempt_count += 1
                            self.actionQueue.remove(proc)
                            proc2.action = Action.enter_CPU
                            proc2.action_enter = self.clock + (self.t_cs >> 1)
                            proc2.action_leave = self.clock + self.t_cs
                            self.ptProc = proc2
                            self.actionQueue.append(proc2)
                            print("time {:d}ms: Process {:s} (tau {:d}ms) will preempt {:s} ".format(
                                self.clock, proc2.name, proc2.tau, proc.name), end="")
                            self.printReady()

                if proc.action == Action.burst and self.clock == proc.action_leave:
                    self.actionQueue.remove(proc)  # pop from actionQueue
                    # end of I/O
                    print("time {:d}ms: Process {:s} ".format(
                        self.clock, proc.name), end="")
                    if len(proc.burst_time) == proc.index + 1:
                        # no burst left
                        proc.action = Action.leave_CPU
                        proc.action_enter = self.clock
                        proc.action_leave = self.clock + (self.t_cs >> 1)
                        proc.remain = -1
                        self.actionQueue.append(proc)
                        print("terminated ", end="")

                    else:
                        print("(tau {:d}ms) ".format(proc.tau), end="")
                        if len(proc.burst_time) - proc.index == 2:
                            # 1 burst left
                            print(
                                "completed a CPU burst; 1 burst to go ", end="")
                        else:
                            # > 1 bursts left
                            print("completed a CPU burst; {:d} bursts to go ".format(
                                len(proc.burst_time) - proc.index - 1), end="")
                        self.printReady()

                        # recalculate tau
                        proc.tau = math.ceil(self.alpha * proc.burst_time[proc.index] +
                                             (1-self.alpha) * proc.tau)
                        print("time {:d}ms: Recalculated tau = {:d}ms for process {:s} ".format(
                            self.clock, proc.tau, proc.name), end="")
                        self.printReady()

                        # starting I/O
                        proc.action = Action.leave_CPU
                        proc.action_enter = self.clock
                        proc.action_leave = self.clock + (self.t_cs >> 1)
                        proc.remain = 0
                        self.actionQueue.append(proc)
                        print("time {:d}ms: Process {:s} switching out of CPU; ".format(
                            self.clock, proc.name), end="")
                        print("will block on I/O until time {:d}ms ".format(
                            proc.action_leave + proc.block_time[proc.index]), end="")
                    self.printReady()

                if proc.action == Action.block and self.clock == proc.action_leave:
                    self.actionQueue.remove(proc)  # pop from actionQueue
                    print("time {:d}ms: Process {:s} (tau {:d}ms) ".format(
                        self.clock, proc.name, proc.tau), end="")
                    proc.index += 1
                    proc.action = Action.ready
                    self.readyQueue.append(proc)
                    self.readyQueue.sort(key=self.__sort_by_remain)

                    estmate = -1
                    if self.btProc != None and self.btProc.action == Action.burst:
                        self.btProc.remain -= self.clock - self.btProc.action_enter
                        estmate = self.btProc.tau - \
                            (self.btProc.burst_time[self.btProc.index] -
                             self.btProc.remain)
                    if estmate < proc.tau:
                        # if self.btProc == None or newT < proc.tau or self.btProc.action != Action.burst:
                        # no preemption
                        proc.action_enter = self.clock
                        proc.action_leave = self.clock + (self.t_cs >> 1)
                        proc.remain = proc.burst_time[proc.index]
                        print("completed I/O; added to ready queue ", end="")
                        self.printReady()
                    else:
                        # if remain burst time is greater than I/O ended
                        #     process burst time, preemption
                        # take bursting process out of CPU, then insert new process
                        self.btProc.action = Action.leave_CPU
                        self.btProc.action_enter = self.clock
                        self.btProc.action_leave = self.clock + \
                            (self.t_cs >> 1)
                        self.actionQueue.remove(self.btProc)
                        proc.action = Action.enter_CPU
                        proc.action_enter = self.clock + (self.t_cs >> 1)
                        proc.action_leave = self.clock + self.t_cs
                        self.ptProc = proc
                        self.ptProc.preempt_count += 1
                        self.actionQueue.append(proc)
                        print("completed I/O; preempting {:s} ".format(self.btProc.name),
                              end="")
                        self.printReady()

            # push new process into CPU
            if self.btProc == None and len(self.readyQueue) > 0:
                self.btProc = self.readyQueue.pop(0)
                self.btProc.action = Action.enter_CPU
                self.btProc.action_enter = self.clock
                self.btProc.action_leave = self.clock + \
                    (self.t_cs >> 1)
                self.actionQueue.append(self.btProc)

            # insert new process
            self.clock += 1

        print("time {:d}ms: Simulator ended for SRT ".format(self.clock - 1), end="")
        self.printReady()

        tmp = [sum(x.burst_time) for x in self.endedQueue]
        avg_bt = sum(tmp) / len(tmp)  # average burst time
        tmp = [x.action_leave - x.arrival_time - sum(x.burst_time) - 
            sum(x.block_time) - 2 * x.switch_count * (self.t_cs >> 1)
            for x in self.endedQueue]
        avg_wt = sum(tmp) / len(tmp)  # average wait time
        tmp = [x.action_leave - x.arrival_time for x in self.endedQueue]
        avg_tt = sum(tmp) / len(tmp)  # average turnaround time
        tmp = [x.switch_count for x in self.endedQueue]
        sum_cs = sum(tmp)             # total context switches
        tmp = [x.preempt_count for x in self.endedQueue]
        sum_pp = sum(tmp)             # total preemptions
        return (avg_bt, avg_wt, avg_tt, sum_cs, sum_pp)


    """
    print readyQueue
    """

    def printReady(self):
        print("[Q", end="")
        if len(self.readyQueue) == 0:
            print(" <empty>", end="")
        else:
            for p in self.readyQueue:
                print(" " + p.name, end="")
        print("]")

    """ 
    sort order
        (a) CPU burst completion
        (b) I/O burst completion(i.e., back to the ready queue)
        (c) new process arrival
        If ties still happens, break tie with ID order
    """

    def __sort_by_action(self, proc: Process):
        if proc.action == Action.burst:
            return ord(proc.name[0]) - 65 + 26 * 3
        elif proc.action == Action.block:
            return ord(proc.name[0]) - 65 + 26 * 2
        elif proc.action == Action.new:
            return ord(proc.name[0]) - 65 + 26 * 1
        else:
            return ord(proc.name[0]) - 65 + 26

    """
    sort order
        remaining time
        If ties still happens, break tie with ID order
    """

    def __sort_by_remain(self, proc: Process):
        return proc.remain + ord(proc.name[0]) - 65

File: SRT.java
import java.util.List;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * A class that represents the First Come First Serve CPU scheduling algorithm
 */
public class SRT extends CPUSchedual {
    public SRT(List<Process> procs, int t_cs, double alpha) {
        readyList = new ArrayList<Process>(procs.size());
        actionList = new ArrayList<Process>(procs.size());
        endedList = new ArrayList<Process>(procs.size());
        fs = new FormatedStdout("SRT", Integer.MAX_VALUE);
        this.t_cs = t_cs;
        this.alpha = alpha;
        clock = 0;

        // deep copy
        Iterator<Process> itr = procs.iterator();
        try {
            while (itr.hasNext()) {
                actionList.add(itr.next().clone());
            }
        } catch (CloneNotSupportedException exc) {
            System.err.print("ERROR: Clone process failed.");
            System.exit(-1);
        }
    }

    @Override
    public double[] simulate() {
        // print all processes
        fs.printAll(actionList);
        System.out.println("time 0ms: Simulator started for SRT [Q <empty>]");
        actionList.sort(new _sortByState());
        // store process that is in load state
        Process btProc = null; // process with LOAD or BURST or UNLOAD state
        while (actionList.size() != 0) {

            // get time from next process
            clock = actionList.get(0).kTime;

            if (clock == 8959) {
                int X = 0;
            }

            while (actionList.size() > 0 && actionList.get(0).kTime == clock) {
                Process proc = actionList.remove(0);
                switch (proc.state) {
                    case NEW:
                        proc.state = States.READY;
                        readyList.add(proc);
                        readyList.sort(new _sortByEstmate());
                        fs.printArrival(clock, proc, readyList);
                        break;

                    case READY:
                        // perpear for load into CPU, load state
                        proc.state = States.LOAD;
                        proc.kTime += t_cs >> 1;
                        actionList.add(proc);
                        btProc = proc;
                        break;

                    case LOAD:
                        // loaded into CPU complete, next state is burst
                        proc.state = States.BURST;
                        proc.kTime += proc.remain;
                        // proc.csCount += 1; // context switch
                        fs.printStartedBurst(clock, proc, readyList);
                        actionList.add(proc);
                        break;

                    case BURST:
                        // burst complete, into unload state
                        proc.state = States.UNLOAD;
                        proc.kTime += t_cs >> 1;
                        proc.remain = 0;
                        if (proc.burstTimes.length - proc.progress == 1) {
                            // process terminated, but unload from CPU first
                            fs.printTerminate(clock, proc, readyList);
                        } else {
                            fs.printEndedBurst(clock, proc, readyList);
                            proc.tau = (int) Math.ceil(alpha * proc.burstTimes[proc.progress] + (1 - alpha) * proc.tau);
                            fs.printReTau(clock, proc, readyList);
                            proc.kTime += proc.blockTimes[proc.progress]; // just for print
                            fs.printStartedBlock(clock, proc, readyList);
                            proc.kTime -= proc.blockTimes[proc.progress];
                        }
                        actionList.add(proc);
                        break;

                    case UNLOAD:
                        // unload from CPU
                        if (proc.remain > 0) {
                            // being preempted
                            proc.state = States.READY;
                            readyList.add(proc);
                            readyList.sort(new _sortByEstmate());
                        } else if (proc.burstTimes.length - proc.progress == 1) {
                            // no more burst to go, process terminate
                            proc.state = States.TERMINATED;
                            endedList.add(proc);
                        } else {
                            // more bursts to go, and not being preempted
                            proc.state = States.BLOCK;
                            proc.kTime += proc.blockTimes[proc.progress];
                            actionList.add(proc);

                        }
                        proc.csCount += 1; // context switch
                        btProc = null;
                        break;

                    case BLOCK:
                        // block complete, no preemption in FCFS
                        proc.state = States.READY;
                        proc.progress += 1;
                        proc.remain = proc.burstTimes[proc.progress];
                        readyList.add(proc);
                        readyList.sort(new _sortByEstmate());
                        // check preemption
                        if (btProc != null && btProc.state == States.BURST) {
                            btProc.remain = btProc.kTime - clock;
                            int estLeft = btProc.tau - btProc.burstTimes[btProc.progress] + btProc.remain;
                            if (estLeft > proc.tau) {
                                // perform preemption
                                btProc.state = States.UNLOAD;
                                btProc.kTime = clock + (t_cs >> 1);
                                fs.printEndedBlock(clock, proc, btProc, readyList);
                            } else {
                                fs.printEndedBlock(clock, proc, null, readyList);
                            }
                        } else {
                            fs.printEndedBlock(clock, proc, null, readyList);
                        }

                        break;

                    default:
                        System.err.println("Error: unexpected state.");
                        System.exit(-1);
                }

                // sort
                actionList.sort(new _sortByState());
            }

            // empty CPU, add new process
            if (readyList.size() > 0) {
                if (btProc == null) {
                    Process tmp = readyList.remove(0);
                    tmp.kTime = clock;
                    actionList.add(tmp);
                    actionList.sort(new _sortByState());
                } else if (btProc != null && btProc.state == States.BURST) {
                    // check if any ready process should preempts bursting process
                    btProc.remain = btProc.kTime - clock;
                    int estLeft = btProc.tau - btProc.burstTimes[btProc.progress] + btProc.remain;
                    Process tmp = readyList.get(0);
                    if (estLeft > tmp.tau) {
                        btProc.state = States.UNLOAD;
                        btProc.kTime = clock + (t_cs >> 1);
                        System.out.printf("time %dms: Process %c (tau %dms) will preempt %c ", clock, tmp.name, tmp.tau,
                                btProc.name);
                        fs.printReady(readyList);
                        actionList.sort(new _sortByState());
                    }
                }
            }
        }

        System.out.printf("time %dms: Simulator ended for SRT ", clock);
        fs.printReady(readyList);

        double retVal[] = new double[5];
        // calRetVal(retVal);
        // System.err.printf("%.3f %.3f %.3f %.0f %.0f", retVal[0], retVal[1],
        // retVal[2], retVal[3], retVal[4]);
        return retVal;
    }
}
